Issues with multiple includes of UseF2Py

Another issue from me, though this one is a bit more basic/CMake-y than #20. 

As said in #20, we are looking at maybe moving to use f2py-cmake in our code. In doing so, I encountered an issue. The circumstances are that we have two repos that would like to use it, say A and B. Repo B is standalone, but repo A includes repo B (as a submodule or whatever). Since we want both A and B to be able to build, both A and B have `include(UseF2Py)` calls in them. However, it seems like `UseF2Py.cmake` is not set up to handle that. My investigations seem to show it is due to (at least):

1. The `_Python` and `_Python3` variables not being `CACHE`
2. The `F2Py::Headers` not being `GLOBAL`

I created a demonstration of this at https://github.com/mathomp4/f2py-cmake-bug which has a case of multiple calls to `include(UseF2Py.cmake)`

In this repo, I used the:
```
pipx run f2py-cmake vendor cmake
```
command to add `UseF2Py.cmake` in both "repos" and then added "fixed" versions.

So, if you build with `-DFIXED_FILE:BOOL=OFF` you'll see the original error:
```
CMake Error at cmake/UseF2Py.cmake:43 (target_link_libraries):
  Target "test_object" links to:

    ::NumPy

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

Call Stack (most recent call first):
  src/CMakeLists.txt:18 (f2py_object_library)


-- Generating done (0.0s)
```

But if you run with `-DFIXED_FILE:BOOL=ON` then it all succeeds.

Note: At this point, all I've shown is this seems to let things build. In the real case where I'm trying to use this, I have no idea how to run any of the code, so perhaps there are more issues, but this at least is something.

I'm hestitant to make a PR with my fix only because, well, I need to do more testing on our end. But this seems to be usable.

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -4,13 +4,13 @@ ci:
 
 repos:
   - repo: https://github.com/adamchainz/blacken-docs
-    rev: "1.16.0"
+    rev: "1.19.1"
     hooks:
       - id: blacken-docs
         additional_dependencies: [black==24.*]
 
   - repo: https://github.com/pre-commit/pre-commit-hooks
-    rev: "v4.6.0"
+    rev: "v5.0.0"
     hooks:
       - id: check-added-large-files
       - id: check-case-conflict
@@ -32,22 +32,22 @@ repos:
       - id: rst-directive-colons
       - id: rst-inline-touching-normal
 
-  - repo: https://github.com/pre-commit/mirrors-prettier
-    rev: "v4.0.0-alpha.8"
+  - repo: https://github.com/rbubley/mirrors-prettier
+    rev: "v3.4.2"
     hooks:
       - id: prettier
         types_or: [yaml, markdown, html, css, scss, javascript, json]
         args: [--prose-wrap=always]
 
   - repo: https://github.com/astral-sh/ruff-pre-commit
-    rev: "v0.5.0"
+    rev: "v0.8.3"
     hooks:
       - id: ruff
         args: ["--fix", "--show-fixes"]
       - id: ruff-format
 
   - repo: https://github.com/pre-commit/mirrors-mypy
-    rev: "v1.10.1"
+    rev: "v1.13.0"
     hooks:
       - id: mypy
         files: src|tests
@@ -76,13 +76,13 @@ repos:
         exclude: .pre-commit-config.yaml
 
   - repo: https://github.com/abravalheri/validate-pyproject
-    rev: "v0.18"
+    rev: "v0.23"
     hooks:
       - id: validate-pyproject
         additional_dependencies: ["validate-pyproject-schema-store[all]"]
 
   - repo: https://github.com/python-jsonschema/check-jsonschema
-    rev: "0.28.6"
+    rev: "0.30.0"
     hooks:
       - id: check-dependabot
       - id: check-github-workflows

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -109,9 +109,6 @@ disallow_untyped_defs = true
 disallow_incomplete_defs = true
 
 
-[tool.ruff]
-src = ["src"]
-
 [tool.ruff.lint]
 extend-select = [
   "B",        # flake8-bugbear
@@ -142,8 +139,6 @@ ignore = [
   "ISC001",   # Conflicts with formatter
 ]
 isort.required-imports = ["from __future__ import annotations"]
-# Uncomment if using a _compat.typing backport
-# typing-modules = ["f2py_cmake._compat.typing"]
 
 [tool.ruff.lint.per-file-ignores]
 "tests/**" = ["T20"]

--- a/src/f2py_cmake/cmake/UseF2Py.cmake
+++ b/src/f2py_cmake/cmake/UseF2Py.cmake
@@ -8,7 +8,7 @@ if(CMAKE_VERSION VERSION_LESS 3.17)
   message(FATAL_ERROR "CMake 3.17+ required")
 endif()
 
-include_guard(GLOBAL)
+include_guard(DIRECTORY)
 
 if(TARGET Python::NumPy)
   set(_Python Python)

--- a/tests/packages/f90dual/CMakeLists.txt
+++ b/tests/packages/f90dual/CMakeLists.txt
@@ -1,0 +1,9 @@
+cmake_minimum_required(VERSION 3.24)
+
+project(test
+  LANGUAGES Fortran C
+)
+
+list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake")
+
+add_subdirectory(src)

--- a/tests/packages/f90dual/src/CMakeLists.txt
+++ b/tests/packages/f90dual/src/CMakeLists.txt
@@ -1,0 +1,19 @@
+add_subdirectory(subrepo)
+
+set (CMAKE_FIND_FRAMEWORK LAST)
+find_package(
+  Python
+  COMPONENTS Interpreter Development.Module NumPy
+  REQUIRED)
+
+include(UseF2Py)
+
+message(STATUS "Building test f2py module")
+
+f2py_object_library(test_object OBJECT)
+f2py_generate_module(test_ test_py.F90 OUTPUT_VARIABLE test_files)
+python_add_library(test_ MODULE "${test_files}" WITH_SOABI)
+target_link_libraries(test_ PRIVATE test_object)
+install(TARGETS test_ DESTINATION lib/Python)
+
+message(STATUS "Done building test f2py module")

--- a/tests/packages/f90dual/src/subrepo/CMakeLists.txt
+++ b/tests/packages/f90dual/src/subrepo/CMakeLists.txt
@@ -1,0 +1,9 @@
+cmake_minimum_required(VERSION 3.24)
+
+project(test-subrepo
+  LANGUAGES Fortran C
+)
+
+list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake")
+
+add_subdirectory(src)

--- a/tests/packages/f90dual/src/subrepo/src/CMakeLists.txt
+++ b/tests/packages/f90dual/src/subrepo/src/CMakeLists.txt
@@ -1,0 +1,18 @@
+set(CMAKE_FIND_FRAMEWORK LAST)
+
+find_package(
+  Python
+  COMPONENTS Interpreter Development.Module NumPy
+  REQUIRED)
+
+include(UseF2Py)
+
+message(STATUS "Building subrepo f2py module")
+
+f2py_object_library(subrepo_object OBJECT)
+f2py_generate_module(subrepo_ subrepo_py.F90 OUTPUT_VARIABLE subrepo_files)
+python_add_library(subrepo_ MODULE "${subrepo_files}" WITH_SOABI)
+target_link_libraries(subrepo_ PRIVATE subrepo_object)
+install(TARGETS subrepo_ DESTINATION lib/Python)
+
+message(STATUS "Done building subrepo f2py module")

--- a/tests/packages/f90dual/src/subrepo/src/subrepo_piy.F90
+++ b/tests/packages/f90dual/src/subrepo/src/subrepo_piy.F90
@@ -1,0 +1,3 @@
+subroutine bobo()
+   print *, "bobo"
+end subroutine bobo

--- a/tests/packages/f90dual/src/test_py.F90
+++ b/tests/packages/f90dual/src/test_py.F90
@@ -1,0 +1,3 @@
+subroutine bobo()
+   print *, "bobo"
+end subroutine bobo

--- a/tests/test_package.py
+++ b/tests/test_package.py
@@ -1,12 +1,16 @@
 from __future__ import annotations
 
 import importlib.metadata
+import shutil
+import subprocess
 import zipfile
 from pathlib import Path
 
+import pytest
 from scikit_build_core.build import build_wheel
 
 import f2py_cmake as m
+import f2py_cmake.vendor
 
 DIR = Path(__file__).parent.resolve()
 
@@ -15,6 +19,7 @@ def test_version():
     assert importlib.metadata.version("f2py_cmake") == m.__version__
 
 
+@pytest.mark.skipif(shutil.which("cmake") is None, reason="CMake not found")
 def test_f77(monkeypatch, tmp_path):
     monkeypatch.chdir(DIR / "packages/f77")
     build_dir = tmp_path / "build"
@@ -30,3 +35,20 @@ def test_f77(monkeypatch, tmp_path):
     build_files = {x.name for x in build_dir.iterdir()}
     assert "fibbymodule.c" in build_files
     assert "fibby-f2pywrappers.f" in build_files
+
+
+def test_f90(monkeypatch, tmp_path):
+    src_dir = tmp_path / "source"
+    build_dir = tmp_path / "build"
+    shutil.copytree(DIR / "packages/f90dual", src_dir)
+    monkeypatch.chdir(src_dir)
+
+    cmake_dir = src_dir / "cmake"
+    cmake_dir.mkdir()
+    f2py_cmake.vendor.vendorize(cmake_dir)
+
+    inner_cmake_dir = src_dir / "src/subrepo/cmake"
+    inner_cmake_dir.mkdir()
+    f2py_cmake.vendor.vendorize(inner_cmake_dir)
+
+    subprocess.run(["cmake", "-S", ".", "-B", str(build_dir)], check=True)

